File: rev1/kanban_app/domain/due_window.py
from datetime import datetime, timedelta

def get_window_start(due_date, scope):
  """Calculate window start based on scope (3h/3d/1w/3mo before due)."""
  try:
    task_due = datetime.strptime(due_date, '%Y-%m-%d').date()
    if scope == 'day':
      return task_due - timedelta(hours=3)
    elif scope == 'week':
      return task_due - timedelta(days=3)
    elif scope == 'month':
      return task_due - timedelta(days=7)
    elif scope == 'year':
      return task_due - timedelta(days=90)
  except:
    pass
  return task_due

def get_due_status(due_date, scope=None):
  """Classify task: noDeadline, scheduled, due, or overdue."""
  if not due_date or not scope:
    return 'noDeadline'
  try:
    task_due = datetime.strptime(due_date, '%Y-%m-%d').date()
    now = datetime.now().date()

    if now > task_due:
      return 'overdue'

    window_start = get_window_start(due_date, scope)
    if now >= window_start:
      return 'due'

    return 'scheduled'
  except:
    return 'noDeadline'

File: rev1/kanban_app/static/js/due_window.js
function parseDate(dateStr) {
  const [year, month, day] = dateStr.split('-');
  return new Date(parseInt(year), parseInt(month) - 1, parseInt(day));
}

function getWindowStart(due_date, scope) {
  const taskDue = parseDate(due_date);
  if (scope === 'day') return new Date(taskDue.getTime() - 3 * 60 * 60 * 1000);
  if (scope === 'week') return new Date(taskDue.getTime() - 3 * 24 * 60 * 60 * 1000);
  if (scope === 'month') return new Date(taskDue.getTime() - 7 * 24 * 60 * 60 * 1000);
  if (scope === 'year') return new Date(taskDue.getTime() - 90 * 24 * 60 * 60 * 1000);
  return taskDue;
}

export function getDueStatus(due_date, scope) {
  if (!due_date || !scope) return 'noDeadline';

  const taskDue = parseDate(due_date);
  const now = new Date();
  now.setHours(0, 0, 0, 0);
  taskDue.setHours(0, 0, 0, 0);

  const currentTime = now.getTime();
  const dueTime = taskDue.getTime();

  if (currentTime > dueTime) return 'overdue';

  const windowStart = getWindowStart(due_date, scope);
  windowStart.setHours(0, 0, 0, 0);

  if (currentTime >= windowStart.getTime()) return 'due';
  return 'scheduled';
}
